homeassistant/components/energyzero/services.py

```diff
@@ -0,0 +1,114 @@
+"""The EnergyZero services."""
+from __future__ import annotations
+
+from datetime import date, datetime
+
+from energyzero import Electricity, Gas, VatOption
+
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import ServiceValidationError
+from homeassistant.util import dt as dt_util
+
+from .const import (
+    ATTR_END,
+    ATTR_INCL_VAT,
+    ATTR_START,
+    DOMAIN,
+    ENERGY_SERVICE_NAME,
+    GAS_SERVICE_NAME,
+    SERVICE_SCHEMA,
+    PriceType,
+)
+from .coordinator import EnergyZeroDataUpdateCoordinator
+
+
+def __get_date(date_input: str | None) -> date | datetime:
+    """Get date."""
+    if not date_input:
+        return dt_util.now().date()
+
+    if value := dt_util.parse_datetime(date_input):
+        return value
+
+    raise ServiceValidationError(
+        translation_domain=DOMAIN,
+        translation_key="invalid_date",
+        translation_placeholders={
+            "date": date_input,
+        },
+    )
+
+
+def __serialize_prices(prices: Electricity | Gas) -> ServiceResponse:
+    """Serialize prices."""
+    return {
+        "prices": [
+            {
+                key: str(value) if isinstance(value, datetime) else value
+                for key, value in timestamp_price.items()
+            }
+            for timestamp_price in prices.timestamp_prices
+        ]
+    }
+
+
+async def __get_prices(
+    coordinator: EnergyZeroDataUpdateCoordinator,
+    call: ServiceCall,
+    price_type: PriceType,
+) -> ServiceResponse:
+    start = __get_date(call.data.get(ATTR_START))
+    end = __get_date(call.data.get(ATTR_END))
+
+    vat = VatOption.INCLUDE if call.data.get(ATTR_INCL_VAT) else VatOption.EXCLUDE
+
+    data: Electricity | Gas
+
+    if price_type == PriceType.GAS:
+        data = await coordinator.energyzero.gas_prices(
+            start_date=start,
+            end_date=end,
+            vat=vat,
+        )
+    else:
+        data = await coordinator.energyzero.energy_prices(
+            start_date=start,
+            end_date=end,
+            vat=vat,
+        )
+
+    return __serialize_prices(data)
+
+
+def register_services(
```

Comment:
```suggestion
async def async_setup_services(
```
I think this function should be async

Reply:
Why? It does not do anything async, only pass async functions around?

Reply:
Or should the `hass.services.async_register` be awaited?

Reply:
The `hass.services.async_register` is not a async function so it cannot be awaited?

![image](https://github.com/home-assistant/core/assets/8398505/e5fa5b1e-1b96-40b1-83b1-682b9d0dac96)

Reply:
No you only should await the `async_setup_services`, it is used somewhat mixed in other integrations. Personally, I'm used to these kinds of functions being async, but maybe someone else can answer that better.

Reply:
I cannot see the benefit of creating an async function that is not doing any awaiting, but sure....

Reply:
Please see https://developers.home-assistant.io/docs/asyncio_categorizing_functions/ for the correct function declaration